cmd/gardener-external-admission-controller/main.go

```diff
@@ -0,0 +1,211 @@
+	var (
+		body           []byte
+		receivedReview = admissionv1beta1.AdmissionReview{}
+	)
+
+	// Read HTTP request body into variable.
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// Verify that the correct content-type header has been sent.
+	if contentType := r.Header.Get("Content-Type"); contentType != "application/json" {
+		err := fmt.Errorf("contentType=%s, expect %s", contentType, "application/json")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	// Deserialize HTTP request body into admissionv1beta1.AdmissionReview object.
+	if _, _, err := p.codecs.Decode(body, nil, &receivedReview); err != nil {
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	// If the request field is empty then do not admit (invalid body).
+	if receivedReview.Request == nil {
+		err := fmt.Errorf("invalid request body (missing admission request)")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+	// If the request does not indicate the correct operation (CREATE,UPDATE) we allow the review without further doing.
+	if receivedReview.Request.Operation != admissionv1beta1.Create && receivedReview.Request.Operation != admissionv1beta1.Update {
+		respond(w, &admissionv1beta1.AdmissionResponse{})
+		return
+	}
+
+	// Check that resource is a persistentvolume
+	persistentVolumeResource := metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumes"}
+	if receivedReview.Request.Resource != persistentVolumeResource {
+		err := fmt.Errorf("invalid object (no persistentvolume)")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	// At this point all validation steps have been passed, let's mutate our object.
+	pv := &corev1.PersistentVolume{}
+	if err := json.Unmarshal(receivedReview.Request.Object.Raw, pv); err != nil {
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	patch := "[]"
+	if isCloudSpecificPersistentVolume(pv) {
+		switch {
+		case pv.Initializers == nil:
+			patch = fmt.Sprintf(`[{"op": "add", "path": "/metadata/initializers", "value": {"pending": [{"name": "%s"}]}}]`, pvLabelInitializerName)
+		case !common.HasInitializer(pv.Initializers, pvLabelInitializerName):
+			patch = fmt.Sprintf(`[{"op": "add", "path": "/metadata/initializers/pending/%d", "value": {"name": "%s"}}]`, len(pv.Initializers.Pending), pvLabelInitializerName)
+		}
+		logger.Infof("Added initializer '%s' for PersistentVolume '%s'", pvLabelInitializerName, pv.Name)
+	}
+
```

Comment:
This could also be in a separate function called admit that returns the admission decision which can be passed to response.